stackslib/src/chainstate/stacks/auth.rs

```diff
@@ -314,6 +315,210 @@ impl MultisigSpendingCondition {
+                TransactionAuthField::PublicKey(ref pubk) => {
+                    if !pubk.compressed() {
+                        have_uncompressed = true;
+                    }
+                }
+            };
+        }
+
+        // must be given the right number of signatures
+        if num_sigs_given != signatures_required {
+            test_debug!(
+                "Failed to deserialize order independent multisig spending condition: got {} sigs, expected {}",
+                num_sigs_given,
+                signatures_required
+            );
+            return Err(codec_error::DeserializeError(format!(
+                "Failed to parse order independent multisig spending condition: got {} sigs, expected {}",
+                num_sigs_given, signatures_required
+            )));
+        }
+
+        // must all be compressed if we're using P2WSH
+        if have_uncompressed && hash_mode == OrderIndependentMultisigHashMode::P2WSH {
+            test_debug!(
+                "Failed to deserialize order independent multisig spending condition: expected compressed keys only"
+            );
+            return Err(codec_error::DeserializeError(
+                "Failed to parse order independent multisig spending condition: expected compressed keys only"
+                    .to_string(),
+            ));
+        }
+
+        Ok(OrderIndependentMultisigSpendingCondition {
+            signer,
+            nonce,
+            tx_fee,
+            hash_mode,
+            fields,
+            signatures_required,
+        })
+    }
+}
+
+impl OrderIndependentMultisigSpendingCondition {
+    pub fn push_signature(
+        &mut self,
+        key_encoding: TransactionPublicKeyEncoding,
+        signature: MessageSignature,
+    ) -> () {
+        self.fields
+            .push(TransactionAuthField::Signature(key_encoding, signature));
+    }
+
+    pub fn push_public_key(&mut self, public_key: StacksPublicKey) -> () {
+        self.fields
+            .push(TransactionAuthField::PublicKey(public_key));
+    }
+
+    pub fn pop_auth_field(&mut self) -> Option<TransactionAuthField> {
+        self.fields.pop()
+    }
+
+    pub fn address_mainnet(&self) -> StacksAddress {
+        StacksAddress {
+            version: C32_ADDRESS_VERSION_MAINNET_MULTISIG,
+            bytes: self.signer.clone(),
+        }
+    }
+
+    pub fn address_testnet(&self) -> StacksAddress {
+        StacksAddress {
+            version: C32_ADDRESS_VERSION_TESTNET_MULTISIG,
+            bytes: self.signer.clone(),
+        }
+    }
+
+    /// Authenticate a spending condition against an initial sighash.
+    /// In doing so, recover all public keys and verify that they hash to the signer
+    /// via the given hash mode.
+    pub fn verify(
+        &self,
+        initial_sighash: &Txid,
+        cond_code: &TransactionAuthFlags,
+    ) -> Result<Txid, net_error> {
+        let mut pubkeys = vec![];
+        let cur_sighash = initial_sighash.clone();
```

Comment:
Let's get rid of `cur_sighash` since it never changes now and just use `initial_sighash`